### PR TITLE
Alerting: Update state manager to accept reserved labels

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
Currently, the scheduler mutates the rule's label and adds a reserved label `grafana_folder` which can override the specified by the user. Also, the state manager "generates" a few reserved labels from the rule: `__alert_rule_namespace_uid__`, `alertname`, `__alert_rule_uid__` that override the labels specified by the user in the rule (as well as result).

This PR unifies propagation and usage of the reserved labels. It updates the state manager to accept extra labels that take precedence over labels specified in the rule and the result. Generation of all reserved labels is consolidated in one place - scheduler. 

**Special notes for your reviewer**:

Also, to make sure that there is no regression, I added tests before introducing changes.

### CODE REVIEW

File: pkg/services/ngalert/state/cache.go [hunk 1]
```diff
@@ -202,20 +228,6 @@ func (c *cache) recordMetrics() {
 	}
 }
 
-// if duplicate labels exist, keep the value from the first set
-func mergeLabels(a, b data.Labels) data.Labels {
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/state/cache.go:206]: Perhaps keep this since we merge labels on Line 112-127?

Comment (grobinson-grafana) [inline pkg/services/ngalert/state/cache.go:206]: Perhaps keep this since we merge labels on Line 112-127?

**Review by grobinson-grafana — APPROVED** [review]
I think it looks OK! I am not sure about extra labels being added in the scheduler, that feels like state management to me, but we can come back to this at a later time.

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/services/ngalert/state/cache.go [hunk 2]
```diff
@@ -33,18 +32,42 @@ func newCache(logger log.Logger, metrics *metrics.State, externalURL *url.URL) *
...
-	labels := result.Instance.Copy()
-	attachRuleLabels(labels, alertRule)
-	ruleLabels, annotations := c.expandRuleLabelsAndAnnotations(ctx, alertRule, labels, result)
+	ruleLabels, annotations := c.expandRuleLabelsAndAnnotations(ctx, alertRule, result, extraLabels)
 
-	// if duplicate labels exist, alertRule label will take precedence
-	lbs := mergeLabels(ruleLabels, result.Instance)
-	attachRuleLabels(lbs, alertRule)
+	lbs := make(data.Labels, len(extraLabels)+len(ruleLabels)+len(result.Instance))
+	dupes := make(data.Labels)
```
Comment (JacobsonMT) [inline pkg/services/ngalert/state/cache.go:42]: NIT: This section could be simplified and reuse code by modifying `mergeLabels` to return `dupes`. Then this section becomes something like:

```
// if duplicate labels exist, reserved label will take precedence
lbs, dupes := mergeLabels(extraLabels, ruleLabels)
if len(dupes) > 0 {
	c.log.Warn("rule declares one or many reserved labels. Those rules labels will be ignored", "labels", dupes)
}
// if duplicate labels exist, reserved or alert rule label will take precedence
lbs, dupes = mergeLabels(lbs, result.Instance)
if len(dupes) > 0 {
	c.log.Warn("evaluation result contains either reserved labels or labels declared in the rules. Those labels from the result will be ignored", "labels", dupes)
}
```

Alternatively (or in addition to), maybe this section that can be extracted into a separate function for ease of reading.

WDYT?

**Review by JacobsonMT — APPROVED** [review]
LGTM with one nitpick

**Pull request merged**